app/controllers/submissions_controller.rb

```diff
@@ -481,41 +483,31 @@ def download
   end
 
   ##
-  # Download all files from groupings with id in +params[:groupings]+ in a .zip file.
+  # Prepare all files from groupings with id in +params[:groupings]+ to be downloaded in a .zip file.
   ##
-  def download_groupings_files
+  def zip_groupings_files
     assignment = Assignment.find(params[:assignment_id])
 
-    ## create the zip name with the user name to have less chance to delete
-    ## a currently downloading file
-    short_id = assignment.short_identifier
-    zip_name = Pathname.new(short_id + '_' + current_user.user_name + '.zip')
-    zip_path = Pathname.new('tmp') + zip_name
-
-    ## delete the old file if it exists
-    File.delete(zip_path) if File.exist?(zip_path)
-
     groupings = Grouping.includes(:group, :current_submission_used).where(id: params[:groupings]&.map(&:to_i))
 
+    zip_path = zipped_grouping_file_name(assignment)
+
     if current_user.ta?
       groupings = groupings.joins(:ta_memberships).where('memberships.user_id': current_user.id)
     end
 
-    Zip::File.open(zip_path, Zip::File::CREATE) do |zip_file|
-      groupings.each do |grouping|
-        revision_id = grouping.current_submission_used&.revision_identifier
-        group_name = grouping.group.repo_name
-        grouping.group.access_repo do |repo|
-          revision = repo.get_revision(revision_id)
-          repo.send_tree_to_zip(assignment.repository_folder, zip_file, zip_name + group_name, revision)
-        rescue Repository::RevisionDoesNotExist
-          next
-        end
-      end
-    end
+    @current_job = DownloadSubmissionsJob.perform_later(groupings.ids, zip_path.to_s, assignment.id)
+    session[:job_id] = @current_job.job_id
+
+    render 'shared/_poll_job.js.erb'
+  end
 
-    ## Send the Zip file
-    send_file zip_path, disposition: 'inline', filename: zip_name.to_s
+  # download a zip file previously prepared by calling the
+  # zip_groupings_files method
+  def download_zipped_file
+    assignment = Assignment.find(params[:assignment_id])
+    zip_path = zipped_grouping_file_name(assignment)
+    send_file zip_path, disposition: 'inline', filename: File.basename(zip_path)
```

Comment:
Would be good to check if `zip_path` exists (not sure what `send_file` does if the path doesn't exist).